File: calculation/Calculation.py
from math import sqrt, sin, cos, pi, atan
import os

class TIntensity:
    def __init__(self, i, tau, phi):
        self._i = i
        self._tau = tau
        self._phi = phi
    
    @property
    def i(self):
        return self._i

    @property
    def tau(self):
        return self._tau

    @property
    def phi(self):
        return self._phi
    
    @i.setter
    def set_i(self, value):
        self._i = value
    
    @tau.setter
    def set_tau(self, value):
        self._tau = value

    @phi.setter
    def set_phi(self, value):
        self._phi = value

    def __repr__(self):
        return f'{self.i}, {self.tau}, {self.phi}'

class TStokesVector:
    def __init__(self, j=0, q=0, u=0, v=0):
        self.J = j
        self.Q = q
        self.U = u
        self.V = v
        self.error_code = -1  

    @property
    def P(self):
        if abs(self.J) < 1e-8:
            return 0.0
        else:
            result = sqrt(self.Q**2 + self.U**2 + self.V**2) / self.J
            if result > 1:
                result = 1
            return result
    
    def calculate(self, I1, I2, I3, I4, WriteLog):
        A = [([0] * 5) for _ in range(4)]
        X = [0] * 4

        def Swap(k, n):
            z = A[k][k]
            i = k
            for j in range(k + 1, n):
                if abs(A[j][k]) > z:
                    z = A[j][k]
                    i = j
            if i > k:
                for j in range(k, n):
                    A[i][j], A[k][j] = A[k][j], A[i][j]
        
        def LogMatrix():
            if not WriteLog:
                return
            LogFile.write('------------------------------------------------------\n')
            for i in range(4):
                for j in range(4):
                    LogFile.write(f'{A[i][j]:9.4f}')
                LogFile.write(f' | {A[i][4]:9.4f}\n')
        
        self.ErrorCode = -1
        self.J = 0
        self.Q = 0
        self.U = 0
        self.V = 0

        if WriteLog:
            LogFile = open('LogPart01.txt', 'w', encoding="utf-8")
            LogFile.writelines(
                ('Расчет параметров вектора Стокса рассеянного излучения\n',
                '------------------------------------------------------\n',
                f'I1 = {I1.i:9.4f}\n',
                f'    tau1 = {I1.tau:9.4f}\n',
                f'    phi1 = {I1.phi:9.4f}\n',
                f'I2 = {I2.i:9.4f}\n',
                f'    tau2 = {I2.tau:9.4f}\n',
                f'    phi2 = {I2.phi:9.4f}\n',
                f'I3 = {I3.i:9.4f}\n',
                f'    tau3 = {I3.tau:9.4f}\n',
                f'    phi3 = {I3.phi:9.4f}\n',
                f'I4 = {I4.i:9.4f}\n',
                f'    tau4 = {I4.tau:9.4f}\n',
                f'    phi4 = {I4.phi:9.4f}\n')
            )

        A[0][0] = 1
        A[1][0] = 1
        A[2][0] = 1
        A[3][0] = 1

        A[0][1] = cos(2 * I1.phi)
        A[1][1] = cos(2 * I2.phi)
        A[2][1] = cos(2 * I3.phi)
        A[3][1] = cos(2 * I4.phi)

        A[0][2] = sin(2 * I1.phi) * cos(I1.tau)
        A[1][2] = sin(2 * I2.phi) * cos(I2.tau)
        A[2][2] = sin(2 * I3.phi) * cos(I3.tau)
        A[3][2] = sin(2 * I4.phi) * cos(I4.tau)

        A[0][3] = sin(2 * I1.phi) * sin(I1.tau)
        A[1][3] = sin(2 * I2.phi) * sin(I2.tau)
        A[2][3] = sin(2 * I3.phi) * sin(I3.tau)
        A[3][3] = sin(2 * I4.phi) * sin(I4.tau)

        A[0][4] = 2 * I1.i
        A[1][4] = 2 * I2.i
        A[2][4] = 2 * I3.i
        A[3][4] = 2 * I4.i

        n = 4
        LogMatrix()

        for k in range(n):
            if abs(A[k][k]) < 1e-10:
                List = [abs(value) for value in A[k]]
                idx = List.index(max(List[k:]))
                del List
                if idx != k:
                    Swap(k, idx)
                else:
                    if abs(A[k][n]) > 1e-8:
                        self.ErrorCode = 2
                    else:
                        self.ErrorCode = 1
                    return I1, I2, I3, I4

            for j in range(k + 1, n):
                r = A[j][k] / A[k][k]
                for i in range(n + 1):
                    A[j][i] -= r * A[k][i]
                # B[j] -= r * B[k]

            LogMatrix()
    
        for k in reversed(range(n)):
            r = 0
            for j in range(k + 1, n):
                g = A[k][j] * X[j]
                r += g
            if A[k][k] == 0:
                self.error_code = -1
                LogFile.close()
                return
            X[k] = (A[k][4] - r) / A[k][k]
        
        self.J = X[0]
        self.Q = X[1]
        self.U = X[2]
        self.V = X[3]

        if WriteLog:
            LogFile.writelines(
                ('------------------------------------------------------\n',
                f'J = {X[0]:9.4f}\n',
                f'Q = {X[1]:9.4f}\n',
                f'U = {X[2]:9.4f}\n',
                f'V = {X[3]:9.4f}\n',
                '------------------------------------------------------\n',
                f'P = {self.P:9.4f}\n',
                '------------------------------------------------------\n')
            )
            LogFile.close()
        
        self.error_code = 0

def SquareRoot(v, index):
    a = sqrt(sqrt(v.real**2 + v.imag**2))
    phi = atan(v.imag/v.real)
    if v.real < 0:
        phi += pi
    if (index < 0) or (index > 1):
        index = 0
    if index == 1:
        phi += pi
    return complex(a*cos(phi), a*sin(phi))

class TStokesNaturalVector(TStokesVector):
    def __init__(self, J=0, Q=0, U=0, V=0):
        super().__init__(J, Q, U, V)

    def calculateNatural(self, vector, nju, phi, WriteLog):
        if WriteLog:            
            LogFile = open('LogPart02.txt', 'w', encoding="utf-8")            
            LogFile.writelines(
                ('Расчет параметров вектора Стокса естественного излучения\n',
                '------------------------------------------------------\n',
                'Вектор Стокса рассеянного излучения\n',
                f'    J = {vector.J:9.4f}\n',
                f'    Q = {vector.Q:9.4f}\n',
                f'    U = {vector.U:9.4f}\n',
                f'    V = {vector.V:9.4f}\n',
                f'    P = {vector.P:9.4f}\n')
            )
        
        nju2 = nju**2
        
        cphi = cos(phi/180.0)
        sphi = sin(phi/180.0)
        cosphi = complex(cphi, 0)
        sinphi = complex(sphi, 0)             

        sinphi2 = sinphi * sinphi
        sn = nju2 - sinphi2

        sq = SquareRoot(sn, 0)      
        if sq.imag > 0:
            sq = SquareRoot(sn, 1)

        ch1 = cosphi - sq
        zn1 = cosphi + sq
        nju2cosphi = nju2 * cosphi
        ch2 = nju2cosphi - sq
        zn2 = nju2cosphi + sq
        r1 = ch1 / zn1
        r2 = ch2 / zn2

        r1_2 = abs(r1) ** 2
        r2_2 = abs(r2) ** 2
        r1r2_ = r1 * r2.conjugate()
        r1r2_2 = abs(r1r2_) ** 2

        self.J = (vector.J * (r1_2 + r2_2) - vector.Q * (r1_2 - r2_2)) / (2 * r1_2 * r2_2)
        self.Q = (vector.Q * (r1_2 + r2_2) - vector.J * (r1_2 - r2_2)) / (2 * r1_2 * r2_2)
        self.U = (vector.U * r1r2_.real + vector.V * r1r2_.imag) / r1r2_2
        self.V = (vector.V * r1r2_.real - vector.U * r1r2_.imag) / r1r2_2
        
        if WriteLog:
            LogFile.writelines(
            ('Исходные параметры\n',
            '------------------------------------------------------\n',
            f' Phi                    = {phi:9.4f} град.\n',
            f' Nju                    = ({nju.real:9.4f}, {nju.imag:9.4f})\n',
            '------------------------------------------------------\n',
            'Промежуточные переменные\n',
            '------------------------------------------------------\n',
            f' Nju^2                  = ({nju2.real:9.4f}, {nju2.imag:9.4f})\n',
            f' Nju^2-sin(Phi)^2       = ({sn.real:9.4f}, {sn.imag:9.4f})\n',
            f' SQRT(Nju^2-sin(Phi)^2) = ({sq.real:9.4f}, {sq.imag:9.4f})\n',
            f' r1                     = ({r1.real:9.4f}, {r1.imag:9.4f})\n',
            f' r2                     = ({r2.real:9.4f}, {r2.imag:9.4f})\n',

            f' |r1|^2                 = {r1_2:9.4f}\n',
            f' |r2|^2                 = {r2_2:9.4f}\n',

            f' r1*_r2                 = {r1r2_.real:9.4f}, {r1r2_.imag:9.4f})\n',
            f' |r1*_r2|^2             = {r1r2_2:9.4f}\n',
            '------------------------------------------------------\n',
            'Вектор Стокса естественного излучения\n',
            f'  J0                    = {self.J:9.4f}\n',
            f'  Q0                    = {self.Q:9.4f}\n',
            f'  U0                    = {self.U:9.4f}\n',
            f'  V0                    = {self.V:9.4f}\n',
            f'  P0                    = {self.P:9.4f}\n')
            )
            LogFile.close()

if __name__ == '__main__':
    pass

File: pas/Calculation.pas
unit Calculation;

interface

uses
  Complex
  ,DataModuleUnit
  ;

type
  TIntesity = class;
  
  TStokesVector = class
      ValueJ: double;
      ValueQ: double;
      ValueU: double;
      ValueV: double;

      ValueP: double;

      ErrorCode: integer;
    public
      procedure Init();

      procedure Calculate(var I1, I2, I3, I4: TIntesity; WriteLog: Boolean);

      function GetJ(): double;
      function GetQ(): double;
      function GetU(): double;
      function GetV(): double;

      function GetP(): double;
  end;

  TStokesNaturalVector = class(TStokesVector)
    public
      procedure CalculateNatural(StokesVector: TStokesVector; WriteLog: Boolean);
  end;

  TIntesity = class
    ValueI:   double;
    ValueTau: double;
    ValuePhi: double;
  public
    procedure Init(aI, aTau, aPhi: double);

    function GetI(): double;
    function GetTau(): double;
    function GetPhi(): double;

    procedure SetI(aValue: double);
    procedure SetTau(aValue: double);
    procedure SetPhi(aValue: double);
  end;

implementation

procedure TIntesity.Init(aI, aTau, aPhi: double);
begin
  SetI(aI);

  SetTau(aTau);

  SetPhi(aPhi);
end;

function TIntesity.GetI(): double;
begin
  Result := ValueI;
end;

function TIntesity.GetTau(): double;
begin
  Result := ValueTau;
end;

function TIntesity.GetPhi(): double;
begin
  Result := ValuePhi;
end;

procedure TIntesity.SetI(aValue: double);
begin
  ValueI := aValue;
end;

procedure TIntesity.SetTau(aValue: double);
begin
  ValueTau := aValue;
end;

procedure TIntesity.SetPhi(aValue: double);
begin
  ValuePhi := aValue;
end;

procedure TStokesVector.Init();
begin
  ValueJ := 0.0;

  ValueQ := 0.0;

  ValueU := 0.0;

  ValueV := 0.0;

  ErrorCode := -1;
end;

function TStokesVector.GetJ(): double;
begin
  Result := ValueJ;
end;

function TStokesVector.GetQ(): double;
begin
  Result := ValueQ;
end;

function TStokesVector.GetU(): double;
begin
  Result := ValueU;
end;

function TStokesVector.GetV(): double;
begin
  Result := ValueV;
end;

function TStokesVector.GetP(): double;
begin
  if abs(GetJ()) < 1e-8 then
    Result := 0.0
  else
    Result := SQRT(GetQ() * GetQ() + GetU() * GetU() + GetV() * GetV()) / GetJ();

  if Result > 1 then
    Result := 1;
end;
                                
procedure TStokesVector.Calculate(var I1, I2, I3, I4: TIntesity; WriteLog: Boolean);
var
  A: array[1..4, 1..5] of double;
  B, X: array[1..4] of double;

  i, j, k, n, z, idx: integer;

  r, g, max: double;

  LogFile: TextFile;

  procedure Swap(k, n:integer);
  var
    i, j: integer;
    z:    double;
  begin
    z := A[k,k];

    i := k;

    for j := k + 1 to n do
      if abs(A[j, k]) > z then
      begin
        z := A[j, k];

        i := j;
      end;

    if i > k then
    begin
      for j := k to n do
      begin
        z := A[i, j];

        A[i, j] := A[k, j];

        A[k, j] := z;
      end;

      z := B[i];

      B[i] := B[k];

      B[k] := z;
    end;
  end;

  procedure LogMatrix;
  var
    i, j: integer;
  begin
    if not WriteLog then
      exit;
      
    Writeln(LogFile, '------------------------------------------------------');

    for i := 1 to 4 do
    begin
      for j := 1 to 4 do
        Write(LogFile, A[i, j]:10:4, ' ');

      Writeln(LogFile, ' | ', B[i]:10:4);
    end;
  end;

begin
  ErrorCode := -1;

  ValueJ := 0;
  ValueQ := 0;
  ValueU := 0;
  ValueV := 0;

  //Формируем протокол расчета
  if WriteLog then
  begin
    AssignFile(LogFile, 'LogPart01.txt');

    Rewrite(LogFile);

  //Выводим исходные параметры
    Writeln(LogFile, 'Расчет параметров вектора Стокса рассеянного излучения');
    Writeln(LogFile, '------------------------------------------------------');
    Writeln(LogFile, 'I1 = ', I1.GetI():10:4);
    Writeln(LogFile, '  tau1 = ', I1.GetTau():10:4);
    Writeln(LogFile, '  phi1 = ', I1.GetPhi():10:4);
    Writeln(LogFile, 'I2 = ', I2.GetI():10:4);
    Writeln(LogFile, '  tau2 = ', I2.GetTau():10:4);
    Writeln(LogFile, '  phi2 = ', I2.GetPhi():10:4);
    Writeln(LogFile, 'I3 = ', I3.GetI():10:4);
    Writeln(LogFile, '  tau3 = ', I3.GetTau():10:4);
    Writeln(LogFile, '  phi3 = ', I3.GetPhi():10:4);
    Writeln(LogFile, 'I4 = ', I4.GetI():10:4);
    Writeln(LogFile, '  tau4 = ', I4.GetTau():10:4);
    Writeln(LogFile, '  phi4 = ', I4.GetPhi():10:4);
  end;

  B[1] := 2 * I1.GetI();
  B[2] := 2 * I2.GetI();
  B[3] := 2 * I3.GetI();
  B[4] := 2 * I4.GetI();

  A[1, 1] := 1;
  A[2, 1] := 1;
  A[3, 1] := 1;
  A[4, 1] := 1;

  A[1, 2] := cos(2*I1.GetPhi());
  A[2, 2] := cos(2*I2.GetPhi());
  A[3, 2] := cos(2*I3.GetPhi());
  A[4, 2] := cos(2*I4.GetPhi());

  A[1, 3] := sin(2*I1.GetPhi()) * cos(I1.GetTau());
  A[2, 3] := sin(2*I2.GetPhi()) * cos(I2.GetTau());
  A[3, 3] := sin(2*I3.GetPhi()) * cos(I3.GetTau());
  A[4, 3] := sin(2*I4.GetPhi()) * cos(I4.GetTau());

  A[1, 4] := sin(2*I1.GetPhi()) * sin(I1.GetTau());
  A[2, 4] := sin(2*I2.GetPhi()) * sin(I2.GetTau());
  A[3, 4] := sin(2*I3.GetPhi()) * sin(I3.GetTau());
  A[4, 4] := sin(2*I4.GetPhi()) * sin(I4.GetTau());

  LogMatrix;

  n := 4;

  for i := 1 to n do
  begin
    A[i, n + 1] := B[i];

    X[i] := 0;
  end;

  for k := 1 to n do //по строкам
  begin
    if abs(A[k, k]) < 1e-10 then
    begin
      //поиск максимального по модулю в столбце k ниже строки k
      max := abs(A[k, k]);

      idx := k;

      for z := k + 1 to n do
        if abs(A[k, z]) > max then
        begin
          idx := z;

          max := abs(A[k, z]);
        end;

      if idx <> k then
        Swap(k, idx)
      else
      begin
        if abs(A[k, n + 1]) > 1e-8 then
          ErrorCode := 2
        else
          ErrorCode := 1;

        exit;
      end;
    end;

    for j := k + 1 to n do //со следующей строки до последней
    begin
      r := A[j, k] / A[k, k]; //рассчитываем коэффициент

      for i := 1 to n + 1 do //по столбцам до последнего
        A[j, i] := A[j, i] - r * A[k, i];

      B[j] := B[j] - r * B[k];
    end;

    LogMatrix;
  end;

  for k := n downto 1 do
  begin
    r := 0;

    for j := k + 1 to n do
    begin
      g := A[k, j] * X[j];

      r := r + g;
    end;

    X[k] := (B[k] - r) / A[k, k];
  end;

  ValueJ := X[1];

  ValueQ := X[2];

  ValueU := X[3];

  ValueV := X[4];

  if WriteLog then
  begin
    Writeln(LogFile, '------------------------------------------------------');
    Writeln(LogFile, 'J  = ', X[1]:10:4);
    Writeln(LogFile, 'Q  = ', X[2]:10:4);
    Writeln(LogFile, 'U  = ', X[3]:10:4);
    Writeln(LogFile, 'V  = ', X[4]:10:4);
    Writeln(LogFile, '------------------------------------------------------');
    Writeln(LogFile, 'P  = ', GetP():10:4);
    Writeln(LogFile, '------------------------------------------------------');

    CloseFile(LogFile);
  end;

  ErrorCode := 0;
end;

procedure TStokesNaturalVector.CalculateNatural(StokesVector: TStokesVector; WriteLog: Boolean);
var
  nju, nju2, cosphi, sinphi, sinphi2,
  sn, sq, ch1, zn1, ch2, zn2,
  nju2cosphi, r1, r2, r1r2_: TComplex;

  r1_2, r2_2, r1r2_2: double;

  LogFile: TextFile;
begin
  //Формируем протокол расчета
  if WriteLog then
  begin
    AssignFile(LogFile, 'LogPart02.txt');

    Rewrite(LogFile);

  //Выводим исходные параметры
    Writeln(LogFile, 'Расчет параметров вектора Стокса естественного излучения');
    Writeln(LogFile, '------------------------------------------------------');
    Writeln(LogFile, 'Вектор Стокса рассеянного излучения');
    Writeln(LogFile, '  J = ', StokesVector.GetJ():10:4);
    Writeln(LogFile, '  Q = ', StokesVector.GetQ():10:4);
    Writeln(LogFile, '  U = ', StokesVector.GetU():10:4);
    Writeln(LogFile, '  V = ', StokesVector.GetV():10:4);
    Writeln(LogFile, '  P = ', StokesVector.GetP():10:4);
  end;
  //расчет r1 и r2
  nju := dmPublic.Nju;

  nju2 := nju.Multiply(nju);

  cosphi := TComplex.Create;

  cosphi.Init( cos( dmPublic.Phi/180 ), 0);

  sinphi := TComplex.Create;

  sinphi.Init( sin( dmPublic.Phi/180 ), 0);

  sinphi2 := sinphi.Multiply(sinphi);

  sn := nju2.Substract(sinphi2);

  sq := sn.SquareRoot(0);

  if sq.Image > 0 then
    sq := sn.SquareRoot(1);

  ch1 := cosphi.Substract(sq);

  zn1 := cosphi.Add(sq);

  nju2cosphi := nju2.Multiply(cosphi);

  ch2 := nju2cosphi.Substract(sq);

  zn2 := nju2cosphi.Add(sq);

  r1 := ch1.Divide(zn1);

  r2 := ch2.Divide(zn2);

  //расчет действительных и мнимых частей
  r1_2 := r1.Module * r1.Module;

  r2_2 := r2.Module * r2.Module;

  r1r2_ := r1.Multiply(r2.Covariant());

  r1r2_2:= r1r2_.Module * r1r2_.Module;

  //расчет компонент вектора
  ValueJ := (StokesVector.GetJ * (r1_2 + r2_2) - StokesVector.GetQ * (r1_2 - r2_2))
    / (2 * r1_2 * r2_2);

  ValueQ := (StokesVector.GetQ * (r1_2 + r2_2) - StokesVector.GetJ * (r1_2 - r2_2))
    / (2 * r1_2 * r2_2);

  ValueU := (StokesVector.GetU * r1r2_.Real + StokesVector.GetV * r1r2_.Image) / r1r2_2;

  ValueV := (StokesVector.GetV * r1r2_.Real - StokesVector.GetU * r1r2_.Image) / r1r2_2;

  //Формируем протокол расчета
  if WriteLog then
  begin
    Writeln(LogFile, 'Исходные параметры');
    Writeln(LogFile, '------------------------------------------------------');
    Writeln(LogFile, ' Phi                    = ', dmPublic.Phi:10:4, ' град.');
    Writeln(LogFile, ' Nju                    = (', nju.Real:10:4, ',', nju.Image:10:4, ')');
    Writeln(LogFile, '------------------------------------------------------');
    Writeln(LogFile, 'Промежуточные переменные');
    Writeln(LogFile, '------------------------------------------------------');
    Writeln(LogFile, ' Nju^2                  = (', nju2.Real:10:4, ',', nju2.Image:10:4, ')');
    Writeln(LogFile, ' Nju^2-sin(Phi)^2       = (', sn.Real:10:4, ',', sn.Image:10:4, ')');
    Writeln(LogFile, ' SQRT(Nju^2-sin(Phi)^2) = (', sq.Real:10:4, ',', sq.Image:10:4, ')');
    Writeln(LogFile, ' r1                     = (', r1.Real:10:4, ',', r1.Image:10:4, ')');
    Writeln(LogFile, ' r2                     = (', r2.Real:10:4, ',', r2.Image:10:4, ')');

    Writeln(LogFile, ' |r1|^2                 = ', r1_2:10:4);
    Writeln(LogFile, ' |r2|^2                 = ', r2_2:10:4);

    Writeln(LogFile, ' r1*_r2                 = (', r1r2_.Real:10:4, ',', r1r2_.Image:10:4, ')');
    Writeln(LogFile, ' |r1*_r2|^2             = ', r1r2_2:10:4);
    Writeln(LogFile, '------------------------------------------------------');
    Writeln(LogFile, 'Вектор Стокса естественного излучения');
    Writeln(LogFile, '  J0                    = ', GetJ():10:4);
    Writeln(LogFile, '  Q0                    = ', GetQ():10:4);
    Writeln(LogFile, '  U0                    = ', GetU():10:4);
    Writeln(LogFile, '  V0                    = ', GetV():10:4);
    Writeln(LogFile, '  P0                    = ', GetP():10:4);
    CloseFile(LogFile);
  end;
end;

end.
